File: backend/app/services/custody_service.py
import uuid
from sqlalchemy.orm import Session
from app.models.custody_event import CustodyEvent
from app.schemas.custody import CustodyEventCreate, CustodyEventResponse
from app.crypto.signatures import sign_manifest

class CustodyService:
    @staticmethod
    def get_events(db: Session, evidence_id: str = None, event: str = None, organization: str = None, search: str = None):
        query = db.query(CustodyEvent)
        if evidence_id:
            query = query.filter(CustodyEvent.evidence_id == evidence_id)
        if event and event != 'ALL':
            query = query.filter(CustodyEvent.event == event)
        if organization and organization != 'ALL':
            query = query.filter(CustodyEvent.organization.ilike(f"%{organization}%"))
        if search:
            search = search.lower()
            query = query.filter(
                (CustodyEvent.event_id.ilike(f"%{search}%")) |
                (CustodyEvent.evidence_id.ilike(f"%{search}%")) |
                (CustodyEvent.actor.ilike(f"%{search}%"))
            )
            
        results = query.order_by(CustodyEvent.sequence_number.desc()).all()
        return [CustodyService._format_response(r) for r in results]

    @staticmethod
    def get_events_by_evidence_id(db: Session, evidence_id: str):
        results = db.query(CustodyEvent).filter(CustodyEvent.evidence_id == evidence_id).order_by(CustodyEvent.sequence_number.desc()).all()
        return [CustodyService._format_response(r) for r in results]

    @staticmethod
    def create_event(db: Session, data: CustodyEventCreate):
        event_id = f"EVT-{str(uuid.uuid4())[:8].upper()}"
        
        # Generate signature
        manifest = {
            "evidence_id": data.evidenceId,
            "event": data.event,
            "actor": data.actor,
            "organization": data.organization,
            "hash": data.hash
        }
        sig = sign_manifest(manifest)
        
        tx_ref = "0x" + str(uuid.uuid4().hex)
        
        new_event = CustodyEvent(
            event_id=event_id,
            evidence_id=data.evidenceId,
            event=data.event,
            actor=data.actor,
            organization=data.organization,
            hash=data.hash,
            notes=data.notes,
            parent_id=data.parentId,
            signature=sig,
            tx_ref=tx_ref
        )
        db.add(new_event)
        db.commit()
        db.refresh(new_event)
        return CustodyService._format_response(new_event)

    @staticmethod
    def _format_response(e: CustodyEvent) -> CustodyEventResponse:
        return CustodyEventResponse(
            eventId=e.event_id,
            evidenceId=e.evidence_id,
            event=e.event,
            actor=e.actor,
            organization=e.organization,
            timestamp=e.timestamp.strftime('%Y-%m-%d %H:%M:%S UTC') if e.timestamp else "",
            hash=e.hash,
            verification=e.verification,
            signature=e.signature,
            txRef=e.tx_ref,
            notes=e.notes,
            parentId=e.parent_id
        )

File: src/services/custodyService.js
import { apiClient, IS_MOCK_FALLBACK } from './api';
import { mockCustodyEvents } from '../mock/custodyEvents';

let custodyEventsState = [...mockCustodyEvents];

export const custodyService = {
  async getEvents(filters = {}) {
    try {
      if (!IS_MOCK_FALLBACK) {
        const response = await apiClient.get('/custody/events', { params: filters });
        return response.data;
      }
    } catch (err) {
      console.warn('[CustodyService] API request failed, falling back to local custody events store:', err);
    }

    return custodyEventsState.filter((item) => {
      if (filters.evidenceId && item.evidenceId.toUpperCase() !== filters.evidenceId.toUpperCase()) {
        return false;
      }
      if (filters.event && filters.event !== 'ALL' && item.event !== filters.event) {
        return false;
      }
      if (filters.organization && filters.organization !== 'ALL' && !item.organization.includes(filters.organization)) {
        return false;
      }
      if (filters.search) {
        const q = filters.search.toLowerCase();
        return (
          item.eventId.toLowerCase().includes(q) ||
          item.evidenceId.toLowerCase().includes(q) ||
          item.actor.toLowerCase().includes(q) ||
          item.event.toLowerCase().includes(q) ||
          (item.notes && item.notes.toLowerCase().includes(q))
        );
      }
      return true;
    });
  },

  async getEventsByEvidenceId(evidenceId) {
    try {
      if (!IS_MOCK_FALLBACK) {
        const response = await apiClient.get(`/custody/events/${evidenceId}`);
        return response.data;
      }
    } catch (err) {
      console.warn('[CustodyService] API request failed, falling back to local custody events store:', err);
    }

    return custodyEventsState.filter((ev) => ev.evidenceId.toUpperCase() === evidenceId.toUpperCase());
  },

  async recordCustodyEvent(eventPayload) {
    try {
      if (!IS_MOCK_FALLBACK) {
        const response = await apiClient.post('/custody/events', eventPayload);
        return response.data;
      }
    } catch (err) {
      console.warn('[CustodyService] API request failed, falling back to local custody events store:', err);
    }

    const newEvent = {
      eventId: `EVT-${Math.floor(1000 + Math.random() * 9000)}`,
      evidenceId: eventPayload.evidenceId,
      event: eventPayload.event,
      actor: eventPayload.actor || 'analyst-current@cyber.org',
      organization: eventPayload.organization || 'Organization B (Cyber Lab)',
      timestamp: new Date().toISOString().replace('T', ' ').substring(0, 19) + ' UTC',
      hash: eventPayload.hash || '8f3a91bc72f4cd2a4e9b671a5c28e930f1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6',
      verification: 'VERIFIED',
      signature: '3045022100' + Array.from({length: 20}, () => Math.floor(Math.random()*16).toString(16)).join('') + '...VALID',
      txRef: '0x' + Array.from({length: 40}, () => Math.floor(Math.random()*16).toString(16)).join(''),
      notes: eventPayload.notes || 'Custody action recorded in decentralized evidence ledger.',
      parentId: eventPayload.parentId || null
    };

    custodyEventsState.unshift(newEvent);
    return newEvent;
  }
};

